Added Band Plan to bottom of FFT

DallasEpperson opened the pull request:
Picked up where #621 left off, with a few changes. 

![image](https://user-images.githubusercontent.com/3578478/75723450-5a747a00-5caa-11ea-830c-304468ae9434.png)


Previous implementation was reading from resources file every time frequency was scrolled, resized, or refocused. Instead, I am reading from user's config directory on application start. I chose to use the user's config directory instead of application resource file because each user may wish to have their own bands configured due to different allocations for different regions, etc.

Band Plan configuration file follows this pattern:
`/home/dallas/.config/gqrx/bandplan.csv`
```
# minFrequency,maxFrequency,mode,step,color,name

# USA Amateur Radio
135700,       137800,       CW,  10, #90009470, 2200m Ham Band
472000,       479000,       CW,  10, #90009470, 630m Ham Band
1800000,      2000000,      CW,  10, #90009470, 160m Ham Band
3500000,      3600000,      CW,  10, #90009470, 80m CW/Data Ham Band
3600000,      4000000,      CW,  10, #90009470, 80m Phone Ham Band
5351500,      5366500,      USB, 10, #90009470, 60m Ham Band
[.. removed for brevity ...]
241000000000, 250000000000, NFM, 10, #90009470, GHz Ham Band

# Radionavigation - Source: https://www.ntia.doc.gov/files/ntia/publications/2003-allochrt.pdf
9000,         14000,        USB, 10, #99AABD26, Radionavigation
90000,        110000,       USB, 10, #99AABD26, Radionavigation
190000,       200000,       USB, 10, #99C05017, Aeronautical Radionavigation
[... removed for brevity ...]
9000000000,   9200000000,   USB, 10, #99C05017, Aeronautical Radionavigation
9200000000,   9300000000,   USB, 10, #994C8E75, Maritime Radionavigation
9300000000,   9500000000,   USB, 10, #99AABD26, Radionavigation
[... removed for brevity ...]
252000000000, 265000000000, USB, 10, #99E9EB7C, Radionavigation Satellite
252000000000, 265000000000, USB, 10, #99AABD26, Radionavigation

# Broadcast
54000000,  72000000,  AM,  10, #902E98BB, Broadcast Television
76000000,  088000000, AM,  10, #902E98BB, Broadcast Television
88000000,  108000000, WFM, 10, #902E98BB, Broadcast Radio
174000000, 216000000, AM,  10, #902E98BB, Broadcast Television
470000000, 512000000, AM,  10, #902E98BB, Broadcast Television
512000000, 608000000, AM,  10, #902E98BB, Broadcast Television
614000000, 698000000, AM,  10, #902E98BB, Broadcast Television
698000000, 763000000, AM,  10, #902E98BB, Broadcast Television

# Other
108000000, 118000000,AM, 5000, #900000FF, Air Band VOR/ILS
118000000, 137000000,AM, 5000, #900000FF, Air Band Voice
26960000,  27410000, AM, 5000, #90FF0000, CB
156000000, 162025000,NFM,12500,#90000080, Marine
225000000, 380000000,NFM,12500,#90FF0000, Military Air
240000000, 270000000,NFM,12500,#90FF0000, Mil Sat
446000000, 446200000,NFM,6250, #9000FF00, PMR446
```
Spaces are trimmed, leading zeros are optional. Colors are in `#AARRBBGG` format.

Currently, the band plan configuration file is not editable by GQRX, but I can work on that if it is desired. 

This should resolve #205, #265.

DallasEpperson commented:
As an aside, this is my first C++ contribution to any project ever, so please give any constructive criticism you have.

kolbma commented:
@DallasEpperson I'd like to buy you a beer or coffee ;-)

Where should the money go?

15 minutes back I've thought I need to work on the old pull request to get it to run with the current master and then I've seen your new pull request.
My idea was also to make it with a bandplan.csv.
So great job. The most difficult is now to find the right colors for my csv file...

kolbma commented:
@DallasEpperson Hi again, would it be possible to change it like the overlay is only for the part above the frequency numbers?!
There is a small UX problem, because mouse input is handled different for the bottom part. It is difficult to see where this part starts/ends, because same color.

DallasEpperson commented:
@kolbma 
> would it be possible to change it like the overlay is only for the part above the frequency numbers?!

Do you mean like this?

![image](https://user-images.githubusercontent.com/3578478/76023167-9e5bbf00-5ef6-11ea-9ad1-3fd9803bf3f2.png)

> There is a small UX problem, because mouse input is handled different for the bottom part. It is difficult to see where this part starts/ends, because same color.

I'm sorry, I don't understand. Can you elaborate?

argilo commented:
This seems to work well.

A couple comments:

1. It would be nice to provide a default bandplan.csv file.
2. The "Enable Band Plan" checkbox is aligned with the bottom of the "FFT Settings" panel. I don't think any other settings are aligned this way:

![gqrx-band-plan](https://user-images.githubusercontent.com/583749/76312475-8777ec80-62a9-11ea-993d-36acb10d58ab.png)

DallasEpperson commented:
@argilo 
> It would be nice to provide a default bandplan.csv file.  

Do you have any suggestions where to find a region-agnostic bandplan? I'd hesitate to ship one with, for example, US-specific bands.

kolbma commented:
I think only international kind makes sense. Airband, Maritime, FM Radio Broadcast, PMR, CB?

Maybe there should be a wiki page to put on some snippets for copy&paste for different countries.

I've maybe 80 lines of detailed HAM bandplan for one country.
30km over the border the bandplan is different. ;-)

ra1nb0w commented:
maybe for region (1, 2 or 3)? In the meantime can you post your `bandplan.csv` to try out your PR?
thank you.

ra1nb0w commented:
thanks @DallasEpperson

alexmiguel2211 commented:
Hi, thanks for the great job. It may be possible to include the option of placing two or more lines on the same frequency. I would be so grateful.

iforte commented:
I don't see the bandplan checkbox.  Can someone point me in the right direction on getting this functionality on GQRX?

ra1nb0w commented:
@iforte you need to download [this patch](https://github.com/csete/gqrx/pull/767); apply it to gqrx source and then build gqrx as stated on the manual.

argilo commented:
I'm interested in continuing work on this. To start, I've rebased this on the current master. The code builds, but I haven't tested it yet.

ra1nb0w commented:
I am maintaining an up-to-date version of this PR on my [experimental branch](https://github.com/ra1nb0w/gqrx/tree/experimental) that it is used on macports.

argilo commented:
I've done a bit of cleanup, and added a default bandplan (supplied by @DallasEpperson). When gqrx starts, it checks whether `bandplan.csv` exists in the config folder. If not, it writes out the default file.

DallasEpperson commented:
I'm so happy to see movement on this! I made the first few commits and then
dropped the project entirely (work has been crazy.)

On Sat, Nov 14, 2020, 12:00 PM Clayton Smith <notifications@github.com>
wrote:

> I've done a bit of cleanup, and added a default bandplan (supplied by
> @DallasEpperson <https://github.com/DallasEpperson>). When gqrx starts,
> it checks whether bandplan.csv exists in the config folder. If not, it
> writes out the default file.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/csete/gqrx/pull/767#issuecomment-727235427>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA3JU3RBJFT6K3XX263WD3TSP2ZT5ANCNFSM4K756DVQ>
> .
>

ra1nb0w commented:
thank you @argilo

Rua commented:
A small suggestion for an improvement here, is to draw a border of some sort on the edges of the band. When there are two adjacent bands with the same colour, that would help see where the dividing line is between them. See attached image for a crude example.
![bands](https://user-images.githubusercontent.com/3090610/103478941-63ead900-4dca-11eb-88aa-a8e9cd2a5095.png)
